syncstorage/storage/sql/__init__.py

```diff
@@ -456,6 +456,54 @@ def set_items(self, session, userid, collection, items):
         return self._touch_collection(session, userid, collectionid)
 
     @with_session
+    def create_transaction(self, session, userid, collection):
+        """Creates a transaction in batch_uploads table"""
+        collectionid = self._get_collection_id(session, collection)
+        batchid = ts2bigint(session.timestamp)
+        params = {
+            "batch": batchid,
+            "userid": userid,
+            "collection": collectionid
+        }
+        session.query("CREATE_TRANSACTION", params)
+        return batchid
+
+    @with_session
+    def append_items_to_transaction(self, session, batchid, userid,
+                                    collection, items):
+        """Inserts items into batch_upload_items"""
+        rows = []
+        for data in items:
+            id_ = data["id"]
+            row = self._prepare_bui_row(session, batchid, id_, data)
+            rows.append(row)
+        defaults = {"modified": ts2bigint(session.timestamp)}
+        session.insert_or_update("batch_upload_items", rows, defaults)
+        return session.timestamp
+
+    @with_session
+    def commit_transaction(self, session, batchid, userid, collection):
+        collectionid = self._get_collection_id(session, collection)
+        params = {
+            "batch": batchid,
+            "userid": userid,
+            "collection": collectionid,
+            "default_ttl": MAX_TTL
+        }
+        session.query("COMMIT_TRANSACTION", params)
```

Comment:
I wonder whether it's worth adding an explicit timer around this new and scary query, so we can monitor it more closely than the mysql slow-query log might allow.